File: apps/frontend/lib/latex/compile.ts
import { execFile } from "node:child_process";
import { mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

export type CompileResult =
  | { ok: true; pdf: Buffer }
  | { ok: false; error: string; log?: string };

/**
 * Compile a .tex string with pdflatex (installed in the Docker image).
 */
export async function compileLatexPdf(texSource: string): Promise<CompileResult> {
  const dir = await mkdtemp(join(tmpdir(), "tailorcv-tex-"));
  const texPath = join(dir, "resume.tex");
  const pdfPath = join(dir, "resume.pdf");
  const logPath = join(dir, "resume.log");

  try {
    await writeFile(texPath, texSource, "utf8");

    try {
      await execFileAsync(
        "pdflatex",
        [
          "-interaction=nonstopmode",
          "-halt-on-error",
          `-output-directory=${dir}`,
          texPath,
        ],
        {
          cwd: dir,
          timeout: 60_000,
          maxBuffer: 8 * 1024 * 1024,
          env: { ...process.env, PATH: process.env.PATH },
        },
      );
    } catch (e) {
      let log = "";
      try {
        log = await readFile(logPath, "utf8");
      } catch {
        /* ignore */
      }
      const msg = e instanceof Error ? e.message : String(e);
      const tail = log.split("\n").slice(-40).join("\n");
      return {
        ok: false,
        error: `pdflatex failed: ${msg}`,
        log: tail || undefined,
      };
    }

    const pdf = await readFile(pdfPath);
    if (pdf.length < 100 || !pdf.subarray(0, 5).equals(Buffer.from("%PDF-"))) {
      return { ok: false, error: "pdflatex produced an invalid PDF" };
    }
    return { ok: true, pdf };
  } finally {
    await rm(dir, { recursive: true, force: true }).catch(() => undefined);
  }
}

export async function latexAvailable(): Promise<boolean> {
  try {
    await execFileAsync("pdflatex", ["-version"], { timeout: 5_000 });
    return true;
  } catch {
    return false;
  }
}

File: apps/backend/services/compile.py
from __future__ import annotations

import io
import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Optional


@dataclass
class CompileResult:
    ok: bool
    pdf: Optional[bytes] = None
    error: Optional[str] = None
    log: Optional[str] = None
    pages: int = 0


_PAGES_IN_LOG = re.compile(
    r"Output written on .+?\((\d+)\s+pages?",
    re.IGNORECASE,
)


def latex_available() -> bool:
    return shutil.which("pdflatex") is not None


def pdf_page_count(pdf: bytes) -> int:
    """Return page count for a PDF blob (0 if unreadable)."""
    if not pdf or len(pdf) < 100:
        return 0
    try:
        import pdfplumber

        with pdfplumber.open(io.BytesIO(pdf)) as doc:
            return len(doc.pages)
    except Exception:
        pass
    # Fallback: count leaf Page objects (exclude /Pages)
    return len(re.findall(rb"/Type\s*/Page(?!\s*s)", pdf))


def _pages_from_log(log: str) -> int:
    m = _PAGES_IN_LOG.search(log or "")
    if not m:
        return 0
    try:
        return int(m.group(1))
    except ValueError:
        return 0


def compile_pdf(tex_source: str) -> CompileResult:
    with tempfile.TemporaryDirectory(prefix="tailorcv-tex-") as d:
        dirp = Path(d)
        tex_path = dirp / "resume.tex"
        pdf_path = dirp / "resume.pdf"
        log_path = dirp / "resume.log"
        tex_path.write_text(tex_source, encoding="utf-8")

        try:
            subprocess.run(
                [
                    "pdflatex",
                    "-interaction=nonstopmode",
                    "-halt-on-error",
                    f"-output-directory={dirp}",
                    str(tex_path),
                ],
                cwd=str(dirp),
                timeout=60,
                check=True,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.STDOUT,
            )
        except subprocess.CalledProcessError as e:
            log = ""
            try:
                log = log_path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                pass
            tail = "\n".join(log.splitlines()[-40:])
            return CompileResult(
                ok=False, error=f"pdflatex failed: {e}", log=tail or None
            )
        except subprocess.TimeoutExpired:
            return CompileResult(ok=False, error="pdflatex timed out")
        except FileNotFoundError:
            return CompileResult(ok=False, error="pdflatex not installed")

        if not pdf_path.exists():
            return CompileResult(ok=False, error="pdflatex produced no PDF")
        pdf = pdf_path.read_bytes()
        if len(pdf) < 100 or not pdf[:5] == b"%PDF-":
            return CompileResult(ok=False, error="pdflatex produced an invalid PDF")

        log = ""
        try:
            log = log_path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            pass
        pages = _pages_from_log(log) or pdf_page_count(pdf)
        return CompileResult(ok=True, pdf=pdf, log=log or None, pages=pages)
